pyomo/gdp/disjunct.py

```diff
@@ -74,28 +76,45 @@ def process(arg):
 
 class _DisjunctData(_BlockData):
 
+    # [ESJ 08/16/2019] TODO: Is this a good idea? (If not, need to change back
+    # in DisjunctionData also)
+    @property
+    def transformation_block(self):
+        return self._transformation_block
+
     def __init__(self, component):
         _BlockData.__init__(self, component)
         self.indicator_var = Var(within=Binary)
+        # pointer to transformation block if this disjunct has been
+        # transformed. None indicates it hasn't been transformed.
+        self._transformation_block = None
 
     def set_value(self, val):
         _indicator_var = self.indicator_var
-        # Remove everything
-        for k in list(getattr(self, '_decl', {})):
-            self.del_component(k)
-        self._ctypes = {}
-        self._decl = {}
-        self._decl_order = []
-        # Now copy over everything from the other block.  If the other
-        # block has an indicator_var, it should override this block's.
-        # Otherwise restore this block's indicator_var.
+        # if _transformation_block is not None we should yell. else call base
+        # class, check that you have indicator_var and transformation block.
+        _transformation_block = self._transformation_block
+        if not _transformation_block is None:
+            # the disjunct has been transformed, so the truth no longer lives
+            # here and this is a bad idea!
+            raise GDP_Error("Attempting to call set_value on an already-"
+                            "transformed disjunct! Since disjunct %s "
+                            "has been transformed, replacing it here will "
+                            "not effect the model." % self.name)
+
+        # Call the base class set_value. (Note we are changing decl_order by
+        # doing things this way: the indicator var will always be last rather
+        # than first)
+        super(_DisjunctData, self).set_value(val)
```

Comment:
Hopefully nothing cute relies on `indicator_var` existing before it's created later. I would argue that it's bad practice anyways, so *shrug*.